handlers.go

```diff
@@ -230,6 +235,39 @@ func (r *resolver) handleApplications() (map[string]string, error) {
 	return addedApplications, nil
 }
 
+func (r *resolver) allowCharmUpgrade(existingApp *Application, bundleApp *charm.ApplicationSpec, bundleArch string) (bool, error) {
+	// This covers most of v1 charm URL changes, everything else below is to
+	// support channels. Charmstore charms allow channels, but bundles were not
+	// aware of them, with the introduction of Charmhub charms, then we do need
+	// to factor in channels.
+	if existingApp.Charm != bundleApp.Charm {
+		return true, nil
+	}
+	if !r.force && existingApp.Channel != bundleApp.Channel {
+		return false, errors.Errorf("upgrades not supported across channels (existing: %q, requested: %q)", existingApp.Channel, bundleApp.Channel)
+	}
+	// No existing revision found, so assume no upgrades are available.
+	if existingApp.Revision == -1 {
+		return false, nil
+	}
+
+	requestedRev := -1
+	if r.revisionGetter != nil {
+		var err error
+		requestedRev, err = r.revisionGetter(bundleApp.Charm, bundleApp.Series, bundleApp.Channel, bundleArch)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+	}
+	if requestedRev > existingApp.Revision {
+		return true, nil
+	} else if requestedRev != -1 && requestedRev < existingApp.Revision {
+		return false, errors.Errorf("downgrades are not currently supported")
```

Comment:
Technically, you could still do that with V1 with the L244 check. Does it make sense to error here but still allow users to `--force`?

Reply:
Charmstore allows downgrades, so I don't want to break that

Reply:
Right, L244 will take care of that for V1. My question was more about whether we should allow `--force` here to allow a forced downgrade for charmhub as well?

Reply:
I'll make a note about charmhub and people can raise a bug about it if people encounter it.